compiler/noirc_evaluator/src/ssa/function_builder/data_bus.rs

```diff
@@ -122,15 +122,20 @@ impl FunctionBuilder {
             }
             Type::Array(typ, len) => {
                 databus.map.insert(value, databus.index);
-                for i in 0..len {
-                    for (subitem_index, subitem_typ) in typ.iter().enumerate() {
-                        let index = i * typ.len() + subitem_index;
-                        // load each element of the array
-                        let index = self
+
+                let mut index = 0;
+                for _i in 0..len {
+                    for subitem_typ in typ.iter() {
+                        // load each element of the array, and add it to the databus
+                        let index_var = self
                             .current_function
                             .dfg
                             .make_constant(FieldElement::from(index as i128), Type::length_type());
-                        let element = self.insert_array_get(value, index, subitem_typ.clone());
+                        let element = self.insert_array_get(value, index_var, subitem_typ.clone());
+                        index += match subitem_typ {
+                            Type::Array(_, _) | Type::Slice(_) => subitem_typ.element_size(),
+                            _ => 1,
+                        };
```

Comment:
Can you replace the catch-all match with the types you're expecting to match on and make the catch-all point to `unreachable!()`?

Reply:
Done